### PR TITLE
cycle detection with node colouring + path printing

### PR DESCRIPTION
**Pull request by spirosmaggioros**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

File: graph/cycle_detection.cpp [hunk 1]
```diff
@@ -0,0 +1,117 @@
+#include <vector>
+#include <algorithm>
+#include <iostream>
```
Comment (Panquesito7) [inline graph/cycle_detection.cpp:3]: Please add a one-line description of what the library/header is for (**see the example below**).

```c++
#include <iostream>    /// for IO operations
#include <cassert>    /// for assert
```

File: graph/cycle_detection.cpp [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+using namespace std;
```
Comment (Panquesito7) [inline graph/cycle_detection.cpp:5]: This is discouraged. Please strictly use `std::` for each of its members.

File: graph/cycle_detection.cpp [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+class Graph {
+ private:
+  int V;//number of Vertices
+  vector<int> *adj;//adjacency list for our graph
```
Comment (Panquesito7) [inline graph/cycle_detection.cpp:10]:
```suggestion
  int V = 0;            ///< number of Vertices
  vector<int> *adj{};     ///<adjacency list for our graph
```

File: graph/cycle_detection.cpp [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        path.push_back(u);
+      }
+      path.push_back(p.second);//we push the last node
+      path.push_back(p.first);//we add the first node to clearly see the cycle
+      reverse(path.begin(), path.end());//we reverse so we can have a better outcome(because its from end to start)
+      return true;//and we return true if there's a cycle in our graph
+    }
+  }
+};
+
```
Comment (Panquesito7) [inline graph/cycle_detection.cpp:93]: Please add self-test cases so we can verify your program works properly (**see the [typical structure of a program](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) as a reference**). Let us know if you need any help here or on our [**Discord server**](https://the-algorithms.com/discord). 😃

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
For more information, check out our [contributing guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md).
Let us know if you need any help here or on our [Discord server](https://the-algorithms.com/discord). 🙂

File: graph/cycle_detection.cpp (hunk 2 above)
Comment (spirosmaggioros) [reply graph/cycle_detection.cpp:5]: Yeah completely forgot about that , i'll change it today

File: graph/cycle_detection.cpp [hunk 5]
```diff
@@ -0,0 +1,115 @@
+#include <vector> //path saving and node colouring
```
Comment (Panquesito7) [inline graph/cycle_detection.cpp:1]: Please describe how the algorithm works, a Wikipedia or any other source for algorithm explanation, what the algorithm does, etc..

File: graph/cycle_detection.cpp [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+#include <iostream> //for IO operations
```
Comment (Panquesito7) [inline graph/cycle_detection.cpp:2]:
```suggestion
#include <vector>   /// for std::vector
#include <iostream> /// for IO operations
```

File: graph/cycle_detection.cpp [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+class Graph {
```
Comment (Panquesito7) [inline graph/cycle_detection.cpp:4]: Missing class description.

File: graph/cycle_detection.cpp [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+ private:
+  int V;//number of Vertices
+  std::vector<int> *adj;//adjacency list for our graph
```
Comment (Panquesito7) [inline graph/cycle_detection.cpp:7]:
```suggestion
  int V = 0;                        ///> number of Vertices
  std::vector<int> *adj{};     ///> adjacency list for our graph
```

File: graph/cycle_detection.cpp [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        path.push_back(u);
+      }
+      path.push_back(p.second);//we push the last node
+      path.push_back(p.first);//we add the first node to clearly see the cycle
+      std::reverse(path.begin(), path.end());//we reverse so we can have a better outcome(because its from end to start)
+      return true;//and we return true if there's a cycle in our graph
+    }
+  }
+};
+
```
Comment (Panquesito7) [inline graph/cycle_detection.cpp:90]: Please add self-test cases so we can verify your program works properly (**see the [typical structure of a program](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) as a reference**). Let us know if you need any help here or on our [**Discord server**](https://the-algorithms.com/discord). 😃

File: graph/cycle_detection.cpp [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+int main(){
```
Comment (Panquesito7) [inline graph/cycle_detection.cpp:91]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
int main(){
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: graph/cycle_detection.cpp [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+  }
+
+ public:
+  /**
+  * @brief Constructor of our graph
+  *
+  * @param n number of Vertices
+  */
+  Graph(int n) { V = n; adj = new std::vector<int>[V]; }
+  /**
```
Comment (krishnacx) [inline graph/cycle_detection.cpp:49]: Better to give empty lines between functions

File: graph/cycle_detection.cpp [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+
+
+class Graph {
+ private:
+  int V;//number of Vertices
+  std::vector<int> *adj;//adjacency list for our graph
+
+  //typical node colouring , if the node is unvisited we color it as NOT_STARTED
+  //if we've started the traversal , we color it as STARTED
+  //if we've finished the traversal we color it as FINISHED
```
Comment (krishnacx) [inline graph/cycle_detection.cpp:15]: //if the node is unvisited, we color it as NOT_STARTED
//if we've finished the traversal, we color it as FINISHED

File: graph/cycle_detection.cpp [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+  enum status { NOT_STARTED, STARTED, FINISHED };
+
+  /**
+  * @brief Typical DFS funtion for our graph
```
Comment (krishnacx) [inline graph/cycle_detection.cpp:19]: Typo: function

File: graph/cycle_detection.cpp [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+  void addEdge(int v, int w) { 
+    adj[v].push_back(w); 
+    //adj[w].push_back(v);
+  }
+  /**
+  * @brief Function that detects if there's a cycle in the graph
+  *
+  * @param path : a vector that we store the path of the cycle
+  */
+  bool cycle(std::vector<int> &path) const {
```
Comment (krishnacx) [inline graph/cycle_detection.cpp:71]: Function name can be better. IsCycle or checkIfCycleExists

File: graph/cycle_detection.cpp [hunk 15]
```diff
@@ -0,0 +1,126 @@
+#include <vector> //for std::vector<T>
+#include <iostream> //for IO operations
+
+// The algorithm is using the colouring technique for finding a cycle in a graph
+// you can see more details here : https://en.wikipedia.org/wiki/Graph_coloring
+
+
+class Graph {
+ private:
+  int V;//number of Vertices
```
Comment (krishnacx) [inline graph/cycle_detection.cpp:10]: Vertices -> vertices

File: graph/cycle_detection.cpp [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+    if(temp){
+        std::cout << "Cycle detected , path is -> ";
+        for(int &x : path){
+            std::cout << x << " ";
+        }
+        std::cout << '\n';
+    }
+    else{
+        std::cout << "No cycle" << '\n';
+    }
```
Comment (krishnacx) [inline graph/cycle_detection.cpp:124]: No return statement at the end.

File: graph/cycle_detection.cpp [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+    * @param u: number of edges
+    * @param path: path saving array
+    */
+    int v , u;
+    std::cout << "Enter the number of vertices , :";
+    std::cin >> v;
+    std::cout << "Enter the number of edges , :";
+    std::cin >> u;
+    Graph g(v);
+    while(u--){
```
Comment (krishnacx) [inline graph/cycle_detection.cpp:109]: Add another console message to keep specify the edges among vertices. Else end user may not know why the console is waiting for ?
`std::cout << "Provide the vertexe nos for edge no: " << u << '\n';`

File: graph/cycle_detection.cpp [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+        int fr , to; std::cin >> fr >> to;
+        g.addEdge(fr , to);
+    }
+    std::vector<int> path;//here we store-if there's a cycle- the path from start-end
```
Comment (krishnacx) [inline graph/cycle_detection.cpp:113]: cycle- ?

**Review by krishnacx — COMMENTED** [review]
Dont we need to specify this algo in DIRECTORY.md

**Pull request closed without merging**